File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        int i = 0;
        int start = 0;
        int end = 0;
        Arrays.sort(nums);
        while(i < nums.length - 1){
            if(i > 0 && nums[i] == nums[i-1]){
                i++;
                continue;
            }
            start = i+1;
            end = nums.length - 1;
            while(start < end){
                List<Integer> curr = new ArrayList<>();
                int sum = nums[i] + nums[start] + nums[end];
                if(sum == 0){
                    curr.add(nums[i]);
                    curr.add(nums[start]);
                    curr.add(nums[end]);
                    result.add(curr);
                    while(start < end && nums[start] == nums[start + 1]){
                        start++;
                    }  
                    while(start < end && nums[end] == nums[end - 1]){
                        end--;
                    }   
                    start++;
                    end--;
                }
                if(sum > 0){
                    end--;
                }
                if(sum < 0){
                    start++;
                }
            }
            i++;
        }
        return result;

    }
}

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = set()
        i = 0
        while i < len(nums):
            j = i + 1
            k = len(nums) - 1
            while j < k:
                curr = nums[i] + nums[j] + nums[k]
                if(curr == 0):
                    result.add((nums[i], nums[j], nums[k]))
                    j = j + 1
                    k = k - 1
                if(curr > 0):
                    k = k - 1
                if(curr < 0):
                    j = j + 1
            i = i + 1
        return list(result)



        